activerecord/lib/active_record/associations/association.rb

```diff
@@ -211,12 +211,18 @@ def foreign_key_present?
         # the kind of the class of the associated objects. Meant to be used as
         # a sanity check when you are about to assign an associated record.
         def raise_on_type_mismatch!(record)
-          unless record.is_a?(reflection.klass) || record.is_a?(reflection.class_name.constantize)
+          unless record.is_a?(reflection.klass) || valid_reloaded_class_for?(record)
             message = "#{reflection.class_name}(##{reflection.klass.object_id}) expected, got #{record.class}(##{record.class.object_id})"
             raise ActiveRecord::AssociationTypeMismatch, message
           end
         end
 
+        def valid_reloaded_class_for?(record)
```

Comment:
@senny do you think this is a valid method name? wasn't so sure about how to call this.